jsonrpcservice/__init__.py

```diff
@@ -0,0 +1,259 @@
+            # response if id is not set.
+            self._notification = True
+
+        if 'method' not in data:
+            raise InvalidRequest('Missing "method" member in request.')
+
+        if not isinstance(data['method'], str):
+            raise InvalidRequest('"method" of request must be a string.')
+
+        self._method = data['method']
+
+        if 'params' in data:
+            if isinstance(data['params'], list):
+                self._args.extend(data['params'])
+            elif isinstance(data['params'], dict):
+                self._kwargs.update(data['params'])
+            else:
+                raise InvalidRequest('"params" of request must be either object or array if present.')
+
+
+class Service(object):
+    def __init__(self):
+        self._methods = {}
+
+    def method(self, method):
+        if callable(method):
+            self.add_method(method.__name__, method)
+
+            return method
+        else:
+            def wrapper(func):
+                self.add_method(method, func)
+                return func
+
+            return wrapper
+
+    def add_method(self, method_name, func):
+        if method_name in self._methods:
+            raise ValueError('Method "{}" already registered.'.format(method_name))
+
+        self._methods[method_name] = func
+
+    def add_methods(self, methods, prefix=''):
+        if isinstance(methods, (list, tuple)):
+            methods = {method.__name__: method for method in methods}
+        elif not isinstance(methods, dict):
+            methods = {method: getattr(methods, method) for method in dir(methods) if not method.startswith('_')}
+
+        for name, method in methods.items():
+            if callable(method):
+                self.add_method(prefix + name, method)
+
+    def get_methods(self):
+        return self._methods.copy()
+
+    def handle_request(self, request):
+        response = None
+        try:
+            method_name, args, kwargs = self.before_call(request, request.method(), request.args(), request.kwargs())
+            response = SuccessResponse(request, self.call_method(request, method_name, args, kwargs))
+        except JsonRpcError as e:
+            response = ErrorResponse(request, e.code, e.message, e.data, exception=e)
+        except Exception as e:
+            response = ErrorResponse(request, -32603, "Internal error.", exception=e)
+        finally:
+            return self.after_call(request, response)
+
+    def before_call(self, request, method_name, args, kwargs):
+        return method_name, args, kwargs
+
+    def after_call(self, request, response):
+        return response
+
+    def call_method(self, request, method_name, args, kwargs):
+        if method_name not in self._methods:
+            raise MethodNotFound('Method "{}" is not defined.'.format(method_name))
+
+        method = self._methods[method_name]
+
+        try:
+            inspect.getcallargs(method, *request.args(), **request.kwargs())
```

Comment:
This function [is deprecated according to the docs](https://docs.python.org/3/library/inspect.html#inspect.getcallargs).

Reply:
What's the alternative?

Reply:
See the end of the linked paragraph. ;)